### PR TITLE
Add abstract base class for channels

### PR DESCRIPTION
**Pull request by namurphy**
This PR will add a subpackage tentatively called `abstractions` which will include an [abstract base class](https://docs.python.org/3/library/abc.html) tentatively called `AbstractChannel`. This PR is so we can iterate upon the design initiated by @wtbarnes.

This PR comes after a discussion about how [aiapy](https://gitlab.com/LMSAL_HUB/aia_hub/aiapy) and [xrtpy](https://github.com/HinodeXRT/xrtpy) can have a common interface for calculating effective areas and temperature response functions.  Having a common interface would be particularly useful for multi-instrument differential emission measures, and for cross-calibration between different instruments. This could potentially also be useful for other instrument packages like [irispy](https://gitlab.com/LMSAL_HUB/iris_hub/irispy-lmsal). 

This idea is a joint effort among @wtbarnes, @joyvelasquez, @nabobalis, @jslavin, and me.  

@nabobalis & @wtbarnes: since you're package maintainers (I think), please feel free to edit this directly.  

We can follow up on this with separate PRs for things like `AbstractSpectralModel`, or other things.

TODOs before merging:

 - [x] Add a docstring for the class, with an example of how to use it.  
 - [x] Add docstrings for each of the abstract methods, once we decide on the interface
 - [x] Decide if `gain` should be part of the interface
 - [x] Add tentative parameters for each of the methods, with a type hint annotation for the appropriate `astropy.time` object? (I *think* we can do this, but that the parameters would essentially be suggestions and not requirements.)  Should we include the spectral model as well?
- [x] Double-check consistency with topic guide in #111 
- [x] Add API docs page for `response`
- [x] Fix changelog
- [ ] Add links to topic guide where appropriate in docstrings once #111 is merged.
- [x] Test this with `aiapy` LM-SAL/aiapy#344 
- [ ] Test this with `xrtpy`

### CODE REVIEW

File: sunkit_instruments/abstractions/channels.py
Comment (namurphy) [inline sunkit_instruments/abstractions/channels.py:25]: Do we want to have `gain` be a required part of the interface?

Comment (nabobalis) [reply sunkit_instruments/abstractions/channels.py:25]: Will there be cases where a gain is a bad idea or impossible to work out?

**Comment by namurphy** [comment]
Here's a tentative checklist for things to do:

 - [ ] Add a docstring for the class, with an example of how to use it.  
 - [ ] Add docstrings for each of the abstract methods, once we decide on the interface
 - [ ] Decide if `gain` should be part of the interface
 - [ ] Add tentative parameters for each of the methods, with a type hint annotation for the appropriate `astropy.time` object? (I *think* we can do this, but that the parameters would essentially be suggestions and not requirements.)  Should we include the spectral model as well?

**Comment by dstansby** [comment]
> @nabobalis also brought up the idea of Python 3.8+ [protocol classes](https://peps.python.org/pep-0544/#defining-a-protocol), which would require making this package Python 3.8+ instead of 3.7+.

It looks like this package is now 3.8+ if you want to take this approach!

File: sunkit_instruments/abstractions/channels.py
Comment (wtbarnes) [reply sunkit_instruments/abstractions/channels.py:25]: Maybe. It could always just be set to 1?

Comment (namurphy) [reply sunkit_instruments/abstractions/channels.py:25]: I am happy to leave this decision up to the people who work with observational data and have a good understanding of what gain actually is. 🙃  Defaulting to 1 sounds quite reasonable though.

Comment (namurphy) [inline sunkit_instruments/abstractions/channels.py:25]:
```suggestion
    def gain(self) -> u.DN / u.photon:
        return 1 * u.DN / u.photon
```
If we default to one, we'd want to remove the `@abstractmethod` decorator.

I'm not sure if we'd want this to be dimensional or dimensionless, though.

Comment (wtbarnes) [reply sunkit_instruments/abstractions/channels.py:25]: The gain typically provides a conversion from electrons to counts (or DN) so I think it should have dimensions.

**Comment by wtbarnes** [comment]
I've made a first pass at a full implementation of the `AbstractChannel` using the definitions of the terms laid out in #111. This still needs work, especially in terms of documenting each method. 

Additionally, I've mixed abstract properties/methods with concrete implementations of several of the methods. I'm not sure if this is the right way to go about doing this? Should this go in a separate `BaseChannel` class that can then be subclassed?

I've also implemented a fairly simple `SourceSpectra` class. It uses `xarray` for the underlying data representation and essentially only exists to provide `Quantity` objects for the different axes. This should also probably get its own file at least.

File: sunkit_instruments/abstractions/channels.py
Comment (wtbarnes) [inline sunkit_instruments/abstractions/channels.py:133]: Somehow I made this inconsistent with what I wrote in the topic guide on the response function vocabulary. This should have units of solid angle per pixel. Per this comment https://github.com/sunpy/sunkit-instruments/pull/111#discussion_r1411162940, this name should also change to something more explicit.

Comment (jslavin) [reply sunkit_instruments/abstractions/channels.py:133]: I'd go with pixel solid angle with units of either steradian/pixel or maybe arcsec**2/pixel. The first would tend to be very small of course. In any case conversions between arcsec**2, arcmin**2 and sr are readily available.

File: sunkit_instruments/abstractions/channels.py
Comment (wtbarnes) [reply sunkit_instruments/abstractions/channels.py]: I think it should be left without any value such that anyone using the abstract class has to make the explicit assumption about what their gain is.

**Comment by wtbarnes** [comment]
This should wait on #98 to be merged such that the appropriate links can be added to the docstrings. This abstract base class should also be tested out in several of the existing instrument packages like `aiapy` or `xrtpy` before being merged.

**Comment by wtbarnes** [comment]
TODOs before merging:

- [ ] Double-check consistency with topic guide in #111 
- [ ] Add API docs page for `abstractions`
- [ ] Test this with `aiapy` and `xrtpy`

**Comment by wtbarnes** [comment]
I'm not the biggest fan of having an `abstractions` subpackage though I'm not sure what else to call it.

Maybe we could have a `response` subpackage (where potentially some actual instrument responses could go) and then have `abstractions.py` under that?

**Comment by nabobalis** [comment]
> Maybe we could have a `response` subpackage (where potentially some actual instrument responses could go) and then have `abstractions.py` under that? 
 
I like this

**Comment by namurphy** [comment]
> I'm not the biggest fan of having an `abstractions` subpackage though I'm not sure what else to call it. It seems odd to have "abstractions" so high up in the namespace though since this is a feature that will be used by developers of other packages, not users.

Now that you mention it, I agree!

> Maybe we could have a `response` subpackage (where potentially some actual instrument responses could go) and then have `abstractions.py` under that?

Sounds great to me!  Did you want to make that change?

**Comment by wtbarnes** [comment]
> Sounds great to me! Did you want to make that change?

Yes, I'm happy to. I'm actively working on the branch at the moment anyway.

**Comment by wtbarnes** [comment]
The oldest deps build is failing because for some reason it is picking up `xarray==0.7.0` which is from 8 years ago! https://github.com/pydata/xarray/releases/tag/v0.7.0

File: pyproject.toml
Comment (nabobalis) [inline pyproject.toml:20]:
```suggestion
  "xarray>=2022.9.0",
```

**Review by hayesla — COMMENTED** [review]
this looks good! 

Just a general comment - what about non-imaging instruments that could also use this type of implementation? 

Also it would be good to have an example added here to on how to use this

File: sunkit_instruments/response/abstractions.py
Comment (wtbarnes) [reply sunkit_instruments/response/abstractions.py]: Don't know why that one got skipped!

File: sunkit_instruments/response/thermal.py
Comment (Cadair) [inline sunkit_instruments/response/thermal.py:153]:
```suggestion
            kwargs={'bounds_error': False, 'fill_value': 0.0},
```

File: sunkit_instruments/response/__init__.py
Comment (nabobalis) [inline sunkit_instruments/response/__init__.py:5]: Nitpick, can we add an all and explicit imports?

File: sunkit_instruments/response/thermal.py
Comment (wtbarnes) [reply sunkit_instruments/response/thermal.py:128]: That's a great question. I would hope that yes it could be extended to non-imaging instruments as well. What would the units be in that case? Also, do you have an example of a non-imaging instrument that computes a temperature response? Is it common to compute temperature responses for GOES XRS?

File: sunkit_instruments/response/__init__.py
Comment (wtbarnes) [reply sunkit_instruments/response/__init__.py]: Done

File: sunkit_instruments/response/thermal.py
Comment (wtbarnes) [reply sunkit_instruments/response/thermal.py]: Oops, yes. Fixed.

**Comment by wtbarnes** [comment]
> Just a general comment - what about non-imaging instruments that could also use this type of implementation?

Yes, I think non-imaging instruments should be able to as well. This should presumably be applicable to any imaging instrument and in the context of just the wavelength-resolved effective area, may also be applicable to spectroscopically-resolved data as well. In the case of the temperature response, this should be applicable to any wavelength-integrated instrument primarily observing optically-thin emission.

> Also it would be good to have an example added here to on how to use this

We could create a semi-contrived example in the docs. We could also point to how this is done in external packages like aiapy or xrtpy (once that actually happens). I see this as primarily a developer tool rather than a user tool so maybe just a silly contrived example is sufficient.

**Comment by wtbarnes** [comment]
We had some discussion about this on the community call today (9/25/24). There is a feeling that this should probably go into the core package (`sunpy`) at some point as the functionality is quite general and it would be more convenient for users to just depend on `sunpy` rather than having to depend on `sunkit-instruments` as well. However, there was some hesitancy about having `xarray` become a core dependency (even an optional one). Keeping this in `sunkit-instruments` for now also means we can iterate on it faster as this is still new functionality and the API may change as it gets used in the wild more frequently.

**Comment by Cadair** [comment]
+1 to getting this into core, I think we could have a discussion about the xarray thing when that happens if needed.

**Comment by nabobalis** [comment]
The plan is to merge this and get an alpha release out we can use to experiment on aiapy and xrtpy.

**Review by nabobalis — APPROVED** [review]

**Pull request merged**
